### PR TITLE
[refactor] Perception refactor continued - CraftAssist + some Locobot

### LINKED ISSUE #564: refactor craftassist perception
## Type of Issue

Select the type of issue:
- [ ] Bug report (to report a bug)
- [ ] Feature request (to request an additional feature)
- [x] Tracker (I am just using this as a tracker)
- [ ] Refactor request
- [ ] Documentation Ask

## Checklist

Use this checklist if you are using this issue as a tracker:

- [x] check all functions used:
  - [x] `shape_transforms.py` 
  - [x] rotation.py
- [x] fix lowlevel imports:
  - [x] search.py
  - [x]  low_level_perception.py
  - [x] heuristic_perception.py
  - [x] voxel_models/subcomponent_classifier
- [ ] fix memory imports:
  - [x] low_level_perception.py
  - [x]  heuristic_perception.py
  - [ ] voxel_models/subcomponent_classifier
- [ ] cleanup:
  - [ ] voxel_models/subcomponent_classifier  
  - [ ] voxel_models folder
- [x] memory uses perception's check_inside right now - remove mem method ?
- [ ] move NLU's output to perceive module
- [x] refactor craftassist agent perceive() rewrite
- [x] add memory methods to craftassist agent perceive
- [x] refactor fake agent's perceive()
- [ ] refactor perception perceive methods :
  - [x] `LowLevelMCPerception`
  - [x] `heuristic_perception.PerceptionWrapper`
  - [ ] `SubcomponentClassifierWrapper`
- [x] add memory methods to fake agent perceive
- [ ] refactor heuristic perception's stuff as methods instead of code repetition from before
- [ ] cleanup loco_mc_agent `perceive()`
- [ ] Add `perceive` to Locobot fake_agent
- [ ] remove `self.memory` from all perception modules (craftssist)
- [ ] documentation of agent's perceive:
  - [ ] ca agent
  - [ ] fake agent
- [ ] documentation of cleaned up perceive methods in perception

### PR DESCRIPTION
**Pull request by kavyasrinet**
# Description

This PR helps disentangle memory and lowlevel imports inside perception.
The main refactor is for `craftassist` although `locobot`'s `FakeAgent` now has an updated `perceive` method.

Overall in addition to disentangling, this PR proposes major design changes where all reads and writes to memory now happen inside the `memory` component (memory has the new methods that do the updates that perception was doing before) and the `agent` facilitates these in the `perceive` loop. The `agent` process receives updates from `perception` and sends them to `memory` to be written / updated.

**NOTE:** I am adding more tests to increase coverage and will be pushing them in to this branch before the merge.

Fixes #564 
The issue in addition, also has the full changeset and tracker.

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [x] Refactor
- [ ] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

# Testing

Tested on devfair, CI  and locally. The main tracker has been all unit tests in `droidlet` and `agents` folders.

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [x] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I have verified that all scripts in `tests/scripts` runs on hardware.

### CODE REVIEW

**Comment by kavyasrinet** [comment]
@soumith and @anuragprat1k - I'd like review on the following files:
```
agents/loco_mc_agent.py
agents/locobot/locobot_agent.py
droidlet/interpreter/robot/tests/fake_agent.py 
droidlet/perception/robot/tests/utils.py
```

The TL;DR of the changes above is: we now have the NLU component as one of the perception modules and the FakeAgent's `perceive()` now tests this. The previous changes to robot's perception weren't being tested as we were still calling `loco_mc_agent`'s `perceive()` in the tests. This PR adds and points to the right implementation of perceive.

**Comment by kavyasrinet** [comment]
@aszlam - I'd like review on the proposed design refactor of perception, memory and agent

File: agents/craftassist/craftassist_agent.py [hunk 1]
```diff
@@ -170,12 +173,15 @@ def init_inventory(self):
 
     def init_memory(self):
         """Intialize the agent memory and logging."""
+        low_level_data = self.low_level_data.copy()
+        low_level_data.update({'check_inside': heuristic_perception.check_inside})
```
Comment (aszlam) [inline agents/craftassist/craftassist_agent.py:177]: kind of nit, but why not just low_level_data["check_inside"] = heuristic_perception.check_inside

File: agents/craftassist/craftassist_agent.py [hunk 2]
```diff
@@ -225,17 +228,38 @@ def init_controller(self):
 
     def perceive(self, force=False):
         """Whenever some blocks are changed, that area will be put into a 
-        buffer which will be force-perceived by the agent in the next step
+        buffer which will be force-perceived by the agent in the next step.
```
Comment (aszlam) [inline agents/craftassist/craftassist_agent.py:231]: can you update the comment block here?  comment makes it seem like everything is about the block clustering etc.

File: agents/craftassist/craftassist_agent.py [hunk 3]
```diff
@@ -225,17 +228,38 @@ def init_controller(self):
...
         Here the agent first clusters areas that are overlapping on the buffer,
-        then run through all perception modules to perceive
-        and finally clear the buffer when perception is done.
+        then run through all perception modules to perceive and finally clears the
+        buffer when perception is done.
+
+        The agent sends all perception updates to memory in order for them to
+        update the memory state.
         """
         self.areas_to_perceive = cluster_areas(self.areas_to_perceive)
```
Comment (aszlam) [inline agents/craftassist/craftassist_agent.py:240]: can you move this down to line 245?

File: agents/craftassist/craftassist_agent.py (hunk 1 above)
Comment (kavyasrinet) [reply agents/craftassist/craftassist_agent.py:177]: Yes yes definitely, updating that :)

File: agents/craftassist/craftassist_agent.py [hunk 4]
```diff
@@ -225,17 +228,38 @@ def init_controller(self):
...
+        # 2. perceive from low_level perception module
+        perception_output = self.perception_modules["low_level"].perceive()
+        self.areas_to_perceive = self.memory.update_with_lowlevel_perception_input(
+            perception_output, self.areas_to_perceive)
+        # 3. perceive from heuristic perception module
+        heuristic_perception_output = {}
+        if force or not self.agent.memory.task_stack_peek():
+            # perceive from heuristic perception module
+            heuristic_perception_output = self.perception_modules["heuristic"].perceive()
+            self.memory.update_with_heuristic_perception_input(heuristic_perception_output)
```
Comment (aszlam) [inline agents/craftassist/craftassist_agent.py:252]: can this just be a self.memory.update(dict_with_all_changes), where dict_with_all_changes = {"blocks" : [...], "triples": [...], ...}

File: agents/craftassist/tests/fake_agent.py [hunk 5]
```diff
@@ -312,11 +314,14 @@ def init_physical_interfaces(self):
 
     def init_memory(self):
         T = FakeMCTime(self.world)
+        low_level_data = self.low_level_data.copy()
+        low_level_data.update({'check_inside': heuristic_perception.check_inside})
```
Comment (aszlam) [inline agents/craftassist/tests/fake_agent.py:318]: same nit as above, why "update"?

File: agents/craftassist/tests/fake_agent.py [hunk 6]
```diff
@@ -479,14 +491,21 @@ def safe_get_changed_blocks(self) -> List[Block]:
     ## World setup
     ######################################
 
-    def set_blocks(self, xyzbms: List[Block], origin: XYZ = (0, 0, 0)):
+    def set_blocks(self, xyzbms: List[Block], boring_blocks: Tuple[int], origin: XYZ = (0, 0, 0)):
```
Comment (aszlam) [inline agents/craftassist/tests/fake_agent.py:494]: as above: could we have a memory.update(dict_with_changes}), and not call e.g. memory.maybe_add_block_to_memory here (but in the memory's .update(data)

File: agents/loco_mc_agent.py [hunk 7]
```diff
@@ -133,8 +131,8 @@ def send_text_command_to_agent(sid, command):
             logical_form = {}
             status = ""
             try:
-                chat_parse = self.chat_parser.get_logical_form(
-                    chat=command, parsing_model=self.chat_parser.parsing_model
+                chat_parse = self.perception_modules["language_understanding"].get_logical_form(
```
Comment (aszlam) [inline agents/loco_mc_agent.py:134]: why is this here?  shouldn't the chat be added, and then this is handled by the normal perceive() call after?

File: droidlet/base_util.py [hunk 8]
```diff
@@ -226,3 +221,177 @@ def to_block_pos(array):
 def to_block_center(array):
     """Return the array centered at [0.5, 0.5, 0.5]"""
     return to_block_pos(array).astype("float") + [0.5, 0.5, 0.5]
+
+
+def astar(agent, target, approx=0, pos="agent"):
```
Comment (aszlam) [inline droidlet/base_util.py:226]: maybe we should put these in a craftassist_util ?

File: droidlet/interpreter/craftassist/tasks.py [hunk 9]
```diff
@@ -406,15 +405,17 @@ def step(self):
             if manhat_dist(agent.pos, target) <= self.DIG_REACH:
                 success = agent.dig(*target)
                 if success:
-                    agent.perception_modules["low_level"].maybe_remove_inst_seg(target)
+                    agent.memory.maybe_remove_inst_seg(target)
```
Comment (aszlam) [inline droidlet/interpreter/craftassist/tasks.py:408]: can't wait to get these out of here... but yeah, not yet

File: agents/craftassist/craftassist_agent.py (hunk 3 above)
Comment (aszlam) [reply agents/craftassist/craftassist_agent.py:240]: sorry I don't understand, self.areas_to_perceive is updated between lines 241 and 244?  if so, can we have whatever update be done in the agent? like returned in the perception_output.  if not, yes, just move to be right before 245 so all mentions are in the same place

File: agents/craftassist/craftassist_agent.py (hunk 4 above)
Comment (aszlam) [reply agents/craftassist/craftassist_agent.py:252]: yes, one method in memory that takes the entire dict of changes as input.   that method in memory can just run each of those separately... but from the agent side imo its better if there is a single method from memory that does all the updates

File: agents/craftassist/tests/fake_agent.py (hunk 6 above)
Comment (aszlam) [reply agents/craftassist/tests/fake_agent.py:494]: I think we could still run all of these in a memory.update(changes) where changes is a dict with the appropriate data

Comment (aszlam) [reply agents/craftassist/tests/fake_agent.py:494]: the memory.update(changes) could ofc choose to not do anything based on the exact changes passed in

File: agents/loco_mc_agent.py (hunk 7 above)
Comment (aszlam) [reply agents/loco_mc_agent.py:134]: yeah but why do we need to get_logical_form again?   like can't we get it from memory?

File: agents/craftassist/craftassist_agent.py (hunk 4 above)
Comment (kavyasrinet) [reply agents/craftassist/craftassist_agent.py:252]: The reason I split them into three sequential steps is because: `heuristic_perception`'s `perceive` works on top of updated `areas_to_perceive` from `low_lelvel`(https://github.com/facebookresearch/fairo/blob/8981edf1112375b37ce60d4b1db439bfbdfa3e1b/droidlet/perception/craftassist/heuristic_perception.py#L535). The update to agent's `areas_to_perceive` now happens inside the memory update after `low_level_perception`(https://github.com/facebookresearch/fairo/blob/ca_perception_refactor/droidlet/memory/craftassist/mc_memory.py#L270)

Comment (aszlam) [reply agents/craftassist/craftassist_agent.py:253]: yeah, I think its fine to split into several steps; but put imo its better if those steps are wrapped in a generalized memory.update(data).  if that update() needs to be run twice thats fine with me.  if we start needing to do clever things to deal with performance we can reconsider or refactor, but I think its better to try to keep the .update() interface uniform and not call the individual methods outside of memory

Comment (kavyasrinet) [reply agents/craftassist/craftassist_agent.py:253]: Right yes, that's what I meant. I will update the function to be one big update (and updates will happen based on the changeset that comes in) but I'll just have to call `update` twice here - interleaving between perception calls

File: agents/loco_mc_agent.py (hunk 7 above)
Comment (kavyasrinet) [reply agents/loco_mc_agent.py:134]: Not sure I follow - in this case we just want to query the parser and return the logical form. What is the alternate route here without having to wait for the logical form to get written to memory ?

Comment (aszlam) [reply agents/loco_mc_agent.py:134]: lol I think we _should_ wait for it to get written to memory!  that way what gets sent to dashboard is definitely whatever the agent used/saw

**Review by aszlam — APPROVED** [review]
